### ISSUE TITLE
Improve memory usage on highly concurrent listing

### ISSUE THREAD

**Issue by rvadim** [minio/minio#12548]
Memory leaks when running rclone backup after updating to the latest version.

## Expected Behavior
Our cluster has 8 nodes, 32 disks, 100TB of data, 3.7M objects. Rclone backsup data in 12 hours. ~ 1600 RPS and consumption of CPU and memory within reasonable limits.
![screenshot-grafana intellij net-2021 06 22-09_12_42](https://user-images.githubusercontent.com/7434399/122880342-3246d980-d33a-11eb-9b76-50f3df2c1322.png)

## Current Behavior
After upgrade from `2020-04-15T19-42-18Z` to `2021-06-07T21-40-51Z` and then to `RELEASE.2021-06-17T00-10-46Z` we have memory consumption issues.
![screenshot-grafana intellij net-2021 06 22-09_13_40](https://user-images.githubusercontent.com/7434399/122880380-3d016e80-d33a-11eb-8ea8-45e212793a77.png)

RPS limited to 400 RPS/IP on the edge Nginx side right now. But during long rclone run it cause OOM after ~2-3 hours.

## Possible Solution
No idea right now.

## Steps to Reproduce (for bugs)
1. Setup distributed minio
2. Fill up with ~1M objects
3. Run rclone 
```
rclone sync minio:${bucket} ${BACKUP_DIR}/${bucket} --config=/etc/rclone/rclone.conf --stats-file-name-length=120 --modify-window=1s --use-server-modtime --delete-during --stats-log-level NOTICE --stats=15s --transfers=8 --checkers=64 --buffer-size 256M --retries=5 --retries-sleep=3s --low-level-retries=5
```

## Context
We need to backup whole data once a day.

## Regression
Actually yes, looks like regression. But I cannot find root cause. Maybe problem in configuration. I tried to chenge `list_quorum` to `disk`, but nothing changed.

## Your Environment
* RELEASE.2021-06-17T00-10-46Z
* Kubernetes custom setup with 2 zones. [archive with yamls: zones.zip](https://github.com/minio/minio/files/6692073/zones.zip)
* Ubuntu 20.04

**Comment by rvadim** [comment]
Also file descriptors:
![screenshot-grafana intellij net-2021 06 22-09_48_08](https://user-images.githubusercontent.com/7434399/122885069-0548f580-d33f-11eb-961c-baf36e20b971.png)

**Comment by rvadim** [comment]
File descriptors:
|![screenshot-grafana intellij net-2021 06 22-09_49_20](https://user-images.githubusercontent.com/7434399/122885188-24478780-d33f-11eb-8ac5-3238dd130feb.png)

**Comment by klauspost** [comment]
@rvadim 

While you are having the excessive memory usage, could you run `mc admin profile start --type=mem,gouroutines myminio&&sleep 10&& mc admin profile stop myminio`. It should generate a `profile.zip`; upload it here.

**Comment by rvadim** [comment]
Hi!
Backup turned off right now, but I catch node `minio-z1-3` with high memory consumption.
![screenshot-grafana intellij net-2021 06 22-13_17_40 (1)](https://user-images.githubusercontent.com/7434399/122916597-59fb6900-d35d-11eb-8202-a78d7548e52a.png)
As you can see, "memory freeing" takes near 1 - 1.5 hours.

`findCache` or it structure may cause problems: https://github.com/minio/minio/blob/master/cmd/metacache-bucket.go#L192 
[profile.zip](https://github.com/minio/minio/files/6693772/profile.zip)

**Comment by klauspost** [comment]
rclone must be issuing a crazy amount of concurrent listing operations.

That said I can't see any usage much above 1G of memory. Which server is the yellow line referring to?

**Comment by rvadim** [comment]
`minio-z1-3`
![profile002](https://user-images.githubusercontent.com/7434399/122921606-f6743a00-d362-11eb-94b8-57f6cbc37f20.png)

**Comment by klauspost** [comment]
@rvadim There is nothing unusual, except the number of listings must be crazy. I see I misspelled `goroutines` above. I suspect `*bucketMetacache.cleanup()` is also trying to clean up the previous listings.

You can set `GOGC=25` to reduce the allocation overhead. Set a lower request limit or reduce the aggressiveness of rclone.

We are working on a revision to the listing code, though this seems like an interesting corner case that I will also consider in the new design.

**Comment by rvadim** [comment]
Switching to `mc mirror` helps to solve the problem. Thank you.

**Comment by harshavardhana** [comment]
Let's keep this open for the new listing changes for comparison? @klauspost

**Comment by klauspost** [comment]
@harshavardhana Sure. This is for a lot, many thousands high rate listings. The memory usage is in maintaining in-memory information about caches until 30 minutes after they have been used. So there are thousands in a short timespan, probably at or above a million.

**Comment by harshavardhana** [comment]
This is fixed by #12605
